src/dlsnode/storage/BucketFile.d

```diff
@@ -808,6 +810,37 @@ public class BucketFile: OutputStream
                 });
     }
 
+    /***************************************************************************
+
+        Reads the header of a record from the current seek position of an open
+        bucket file. The seek position is moved ready to read the record's
+        value.
+
+        Params:
+            waiting_context = ContextAwaitingJob which will
+                wake up the request when the node has finished reading required
+                data.
+            header = (output) header of next record
+            promise = filled with the promise that this instance is doing the
+                      non-blocking read and that the data will be available
+                      upon resuming the caller's context, if not yet available,
+                      or with data read, if available.
+
+        Returns:
+            true if the end of the bucket file was reached
+
+    ***************************************************************************/
+
+    public bool nextRecord ( ContextAwaitingJob waiting_context,
+            ref RecordHeader header, ref Promise* promise )
```

Comment:
A reference to a pointer?

Reply:
Yes. The idea is that the promise instance is owned by the `BucketFile`, and only borrowed to the user by pointer - user passes the pointer (possibly `null` - indicating nothing was promised), and then, if the promise is made, `BucketFile` fills the user's pointer to point to this promise.

I did this to ensure the uniqueness of the promise (you just compare pointers). However, as you point out, that might be unnatural, and I could pass promise by value (filled via `ref`), and ensure uniqueness in some other way.